capa/features/extractors/binja/helpers.py

```diff
@@ -0,0 +1,50 @@
+# Copyright (C) 2020 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at: [package root]/LICENSE.txt
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import re
+from typing import List, Callable
+from dataclasses import dataclass
+
+from binaryninja import LowLevelILInstruction
+from binaryninja.architecture import InstructionTextToken
+
+
+@dataclass
+class DisassemblyInstruction:
+    address: int
+    length: int
+    text: List[InstructionTextToken]
+
+
+LLIL_VISITOR = Callable[[LowLevelILInstruction, LowLevelILInstruction, int], bool]
+
+
+def visit_llil_exprs(il: LowLevelILInstruction, func: LLIL_VISITOR):
+    # BN does not really support operand index at the disassembly level, so use the LLIL operand index as a substitute.
+    # Note, this is NOT always guaranteed to be the same as disassembly operand.
+    for i, op in enumerate(il.operands):
+        if isinstance(op, LowLevelILInstruction) and func(op, il, i):
+            visit_llil_exprs(op, func)
+
+
+def unmangle_c_name(name: str) -> str:
+    # https://learn.microsoft.com/en-us/cpp/build/reference/decorated-names?view=msvc-170#FormatC
+    # Possible variations for BaseThreadInitThunk:
+    # @BaseThreadInitThunk@12
+    # _BaseThreadInitThunk
+    # _BaseThreadInitThunk@12
+    # It is also possible for a function to have a `Stub` appended to its name:
+    # _lstrlenWStub@4
+
+    # A small optimization to avoid running the regex too many times
+    # TODO: this still increases the unit test execution time from 170s to 200s, should be able to accelerate it
+    if name[0] in ["@", "_"]:
+        match = re.match(r"^[@|_](.*?)(Stub)?(@\d+)?$", name)
+        if match:
+            return match.group(1)
```

Comment:
oh, neat, i didn't realize you did the demangling here.

do you think this logic applies to other backends? or is this relevant only to names that BN can pull out?

Reply:
The story is like this, when I run the unit test with BN, it initially fails on these:

```
        # file/exports
        ("kernel32", "file", capa.features.file.Export("BaseThreadInitThunk"), True),
        ("kernel32", "file", capa.features.file.Export("lstrlenW"), True),
```

I checked and the reason is that Binja automatically applies PDB and the function name in the PDB is `@BaseThreadInitThunk@12`. If I disable loading PDB, then these tests pass. So I wrote this small C unmangler to fix the names. 

As for IDA, since it is passing the unit test without issue, there are two possibilities:

1. It does not load the PDB by default, so the name is not changed at all
2. It does load the PDB, and has take care of the name unmangling behind the scene

In either case, this code is not needed

For viv, does it support loading PDB?

Reply:
Ah, ok, I wondered if this had to do with PDB handling.

Correct, both IDA and viv don't load the PDB by default. It might be possible to do manually, but we haven't encountered this edge case yet. Makes sense to keep this in the BN code for now, unless it becomes a wider issue.

Thanks for the explanation!

Reply:
Well there is a case that this becomes a problem for IDA. That when people use capa in IDA, and applied PDB beforehand.

Still, I do not see a lot of value to run capa on things that have PDB, which are predominantly binaries from Microsoft.